Introduce DSP subsystem

yperess opened the pull request:
Introduce a shim layer for platforms that do not have official support
in cmsis-dsp via a custom "core" that exposed to the cmsis-dsp library
by adding the "cmsis_compiler.h" file to the include path.

Commit:
math: Introduce a DSP basicmath subsystem with a cmsis backend

Introduce an API mirroring the CMSIS-DSP's basicmath. If CMSIS_DSP is
enabled, then it will by default be used as a backend. Developers may
opt into a custom backend by setting CONFIG_DSP_BACKEND_CMSIS=n. If
done, the application must provide `zdsp_backend/dsp.h` and optionally
implement the functions in its own .c files.

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
Note: I didn't implement mul/div yet because I wanted to make sure we're on the same page.

stephanosio commented on subsys/math/cmsis_backend/public_overrides/dsp_backend/dsp.h:
```diff
@@ -0,0 +1,99 @@
+
+#include <arm_math.h>
+
+typedef int8_t q1_7_t;
+typedef int16_t q1_15_t;
+typedef int32_t q1_31_t;
+typedef int32_t q9_23_t;
+typedef int32_t q16_16_t;
+typedef int32_t q24_8_t;
```
> q1_7_t: 1 signed integer bit, 7 decimal precision bits
q1_15_t: 1 signed integer bit, 15 decimal precision bits
q1_31_t: 1 signed integer bit, 31 decimal precision bits

You are mapping these types to CMSIS-DSP `q7_t`, `q15_t` and `q31_t`. For CMSIS-DSP, these are all 0 integer bit (ARM Q notation includes sign bit in the "integer" part; would be Q0.31 in the TI notation), so its range is from -1 to +1.

> typedef int32_t q9_23_t;
typedef int32_t q16_16_t;
typedef int32_t q24_8_t;

This type system does not really make much sense to me.

No other math/DSP library I have seen does this, and they do not need to because there is no fundamental difference in the operations of, for instance, Q1.31 (`q31_t` in CMSIS-DSP) and the proposed Q9.23 (`q9_23_t`).

What matters is how the inputs and outputs are interpreted (scaled). You can feed Q9.23 to Q1.31 functions, as long as you interpret (scale) the outputs as such.

Moreover, if we decide to go down this path, there will be no end to this (e.g. why do we support `q9_23_t` and not `q10_22_t`, are we doomed to implement the functions for every single variant out there? -- that will be tens of thousands of functions).

`q7_t`, `q15_t`, `q31_t` for fixed point types should be enough.

stephanosio reviewed (changes requested):
This does not seem to be what we settled on in the [chat](https://discord.com/channels/720317445772017664/1025298281540100126/1027663272515735653):

> the idea is basically to define "Zephyr DSP API" [...] that is 1-to-1 mapping to CMSIS-DSP.

> why do we have to reinvent the API when there is CMSIS-DSP?

> > We should not be using architecture specific libraries beyond the architecture domain itself.
>
> it is not architecture specific. as I already explained, the library in itself provides pure C implementations of all functions, so it can run on any architectures as is.

> let's take a look at "Basic Math" functions for instance:
https://github.com/ARM-software/CMSIS-DSP/blob/main/Include/dsp/basic_math_functions.h
there really is nothing ARM-specific in the "API" functions.

> https://github.com/misaleh/CMSIS-DSP-PULPino
> https://github.com/Nuclei-Software/NMSIS/tree/master/NMSIS/DSP
> examples of other RISC-V vendors forking CMSIS-DSP for their use
> just proves my point that this is the most natural way forward.

> and we do have a full testsuite for all the CMSIS-DSP functions so verification is going to be less of a hassle.

So, unless you plan to document all these functions and add full comprehensive test suites for all of them, please keep the API identical to that of the CMSIS-DSP.

Also, remember that the functions you have implemented here is just a subset of the "Basic Math" functions -- there are hundreds more in the CMSIS-DSP and deviating from the CMSIS-DSP is not going to be a sustainable approach unless we have someone working on this full time.

yperess commented:
Updated to use CMSIS-DSP directly via a shimming "core". Note that this will fail to build on some platforms until https://github.com/zephyrproject-rtos/cmsis/pull/20 is merged.

stephanosio commented on subsys/math/public/cmsis_compiler.h:
```diff
@@ -0,0 +1,65 @@
+ * @param ARG2 Bit position to saturate to (1..32)
+ * @return Saturated value
+ */
+#define __SSAT(ARG1, ARG2)                                                                         \
```
This needs to be defined only for aarch64.

yperess replied:
Currently, with the `HAS_CMSIS_CORE` as is, AARCH32 will select CoreA or CoreR from the CMSIS module. Can you explain in what case do we not need this?

stephanosio replied:
> Currently, with the `HAS_CMSIS_CORE` as is, AARCH32 will select CoreA or CoreR from the CMSIS module. Can you explain in what case do we not need this?

non-ARM architectures?

Also for AArch64, [the CMSIS-DSP doc says you should define `__GNUC_PYTHON__`](https://github.com/ARM-software/CMSIS-DSP#how-to-build-for-aarch64), in which case this header file probably will not be necessary at all.

The same should apply to other non-ARM architectures without CMSIS-Core.

stephanosio replied:
FYI, https://github.com/zephyrproject-rtos/cmsis/blob/093de61c2a7d12dc9253daf8692f61f793a9254a/CMSIS/DSP/CMakeLists.txt#L12

yperess replied:
Thanks, that works. I'm getting one issue with `qemu_nios2` giving me `SKIPPED (FLASH overflow)`. Can you point me at an example of extending the flash space on the qemu?

stephanosio commented on modules/Kconfig.cmsis:
```diff
@@ -22,7 +22,7 @@ endif
 menuconfig CMSIS_DSP
 	bool "CMSIS-DSP Library Support"
-	depends on (CPU_CORTEX && NEWLIB_LIBC) || ARCH_POSIX
+	depends on NEWLIB_LIBC || ARCH_POSIX
```
We can only do this when all CMSIS-DSP functions (i.e. not just "Basic Math") are verified to be working on non-Cortex.

yperess replied:
Is there no incremental approach to this? I really don't think we need to go all in right away do we?

stephanosio replied:
well, one way to do that would be to add `depends on CPU_CORTEX` to all the other Kconfigs (i.e. other than "Basic Math") in the meantime.

yperess replied:
Done

stephanosio commented on subsys/math/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2022 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_CMSIS_DSP)
+  if(NOT CONFIG_HAS_CMSIS_CORE)
+    zephyr_include_directories(public)
+  endif()
+endif()
```
What about dsp.h (with the zdsp_ function definitions identical to CMSIS-DSP)  multiple backends support?

yperess replied:
I think I might be confused. Your comment at https://github.com/zephyrproject-rtos/zephyr/pull/51059#pullrequestreview-1133991645 made it seem like you didn't want to have our own custom wrappers.

> So, unless you plan to document all these functions and add full comprehensive test suites for all of them, please keep the API identical to that of the CMSIS-DSP.
>
> Also, remember that the functions you have implemented here is just a subset of the "Basic Math" functions -- there are hundreds more in the CMSIS-DSP and deviating from the CMSIS-DSP is not going to be a sustainable approach unless we have someone working on this full time.

Does this mean that you would like to have our own functions with CMSIS-DSP as a backend?

stephanosio replied:
> I think I might be confused. Your comment at [#51059 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/51059#pullrequestreview-1133991645) made it seem like you didn't want to have our own custom wrappers.

All I said was to "keep the API identical to that of the CMSIS-DSP", not to get rid of the Zephyr DSP API layer altogether.

> Does this mean that you would like to have our own functions with CMSIS-DSP as a backend?

Yes, at least, that was the compromise and what we settled on in the chat:

> the idea (discussed above) is to define Zephyr DSP API that is a direct 1-to-1 mapping to CMSIS-DSP
> so it will be basically CMSIS-DSP API, but with zdsp_ prefix.

stephanosio commented:
> Updated to use CMSIS-DSP directly via a shimming "core". Note that this will fail to build on some platforms until https://github.com/zephyrproject-rtos/cmsis/pull/20 is merged.

Please pull the CMSIS-DSP PR here then.

ruuddw commented:
I'm confused on the approach, what I recall from the TSC discussion was that if we want to standardize on some DSP API in Zephyr, it should be an architecture neutral solution supported and accepted by the different architectures stakeholders.

Here's some input from my DSP expert colleagues on the ARC dsplib solution. In general they agree to a CMSIS-DSP like/compatible API, that's what we did for the ARC dsplib as well. Here are some examples of that.
Features covered:

- Vector math (abs, add, sub, mult, dot_product, neg, offset, scale, min, max, sine, cosine, sqrt, inv-sqrt, exp, log2, ln, lg)
- Complex vector math (conjugate, dot_product, magnitude, magnitude_squared, compex_mul_complex, complex_mul_real)
- Matrix functions (add, sub, mult, scale, complex add/sub/mult/scale), transpose)
- "FIR like" kernels (real FIR, complex FIR, real convolution, correlation)
- Transforms (CFFT, iCFFT, RFFT, iRFFT)

For most cores the library is offered for a number of datatypes, depending on their support on the particular processor. Typically those include fixed-point (Q31, Q16 usually, sometimes Q7), floating point (f32).
Many functions are offered for complex and real data.

The API is straightforward, to give a few examples:

```
//vector dot product
void _DSPLIB_XYCC dsp_dot_prod_f32(const __AGU f32_t * __restrict pSrcA, const __AGU f32_t * __restrict pSrcB, uint32_t nSamples, f32_t *result);
void _DSPLIB_XYCC dsp_dot_prod_q31(const __AGU q31_t * __restrict pSrcA, const __AGU q31_t * __restrict pSrcB, uint32_t nSamples, q63_t *result);
void _DSPLIB_XYCC dsp_dot_prod_q15(const __AGU q15_t * __restrict pSrcA, const __AGU q15_t * __restrict pSrcB, uint32_t nSamples, q63_t *result);

//addition of 2 vectors
void _DSPLIB_XYCC dsp_add_f32(const __AGU f32_t * __restrict pSrcA, const __AGU f32_t * __restrict pSrcB, __AGU f32_t * __restrict pDst, uint32_t nSamples);
void _DSPLIB_XYCC dsp_add_q31(const __AGU q31_t * __restrict pSrcA, const __AGU q31_t * __restrict pSrcB, __AGU q31_t * __restrict pDst, uint32_t nSamples);
void _DSPLIB_XYCC dsp_add_q15(const __AGU q15_t * __restrict pSrcA, const __AGU q15_t * __restrict pSrcB, __AGU q15_t * __restrict pDst, uint32_t nSamples);
void _DSPLIB_XYCC dsp_add_q7(const __AGU q7_t * __restrict pSrcA, const __AGU q7_t * __restrict pSrcB, __AGU q7_t * __restrict pDst, uint32_t nSamples);

//inverse of square matrix [M x M] = [M x M]
dsp_status _DSPLIB_XYCC dsp_mat_inverse_f32(const matrix_f32_t  *pSrc, matrix_f32_t  *pDst);

// Some flavors of FFT:
dsp_status _DSPLIB_XYCC dsp_cfft_init_f32(cfft_f32_t *handle, uint32_t length, __AGU f32_t * pState);
dsp_status _DSPLIB_XYCC dsp_cfft_init_q31(cfft_q31_t *handle, uint32_t length, __AGU uint32_t * pState);
dsp_status _DSPLIB_XYCC dsp_cfft_init_q15(cfft_q15_t *handle, uint32_t length, __AGU uint32_t * pState);
void dsp_cfft_delete_f32(cfft_f32_t *handle);
void dsp_cfft_delete_q31(cfft_q31_t *handle);
void dsp_cfft_delete_q15(cfft_q15_t *handle);
void _DSPLIB_XYCC dsp_cfft_f32(const cfft_f32_t *handle, const __AGU f32_t * input, __AGU f32_t * output);
void _DSPLIB_XYCC dsp_cfft_q31(const cfft_q31_t *handle, const __AGU q31_t * input, __AGU q31_t * output);
void _DSPLIB_XYCC dsp_cfft_q15(const cfft_q15_t *handle, const __AGU q15_t * input, __AGU q15_t * output);
```

You see we use a neutral 'dsp' prefix, nothing architecture specific. And we have some extra type attributes/compiler hints to support ARC CPU special optimizations supported by the Metaware toolchain. These deal with special DSP XY memories (_DSPLIB_XYCC) and use of automatic load/store of operands ( __AGU).

yperess commented:
@ruuddw that's what I originally proposed in this PR but was told that it doesn't scale well. I'm happy to add it back in. Basically what it did was add a `zephyr/dsp.h` that had all the functions with the `zdsp_` prefix. The subsystem then had an optional backend that you can set to cmsis which just remapped these to the cmsis functions. Applications are then free to define their own backends. Would you like me to bring that back?

stephanosio commented:
> that's what I originally proposed in this PR but was told that it doesn't scale well.

What you had was Zephyr's own DSP API loosely based on CMSIS-DSP (i.e. not 1-to-1 mapping to CMSIS-DSP) and that was the criticism.

> The subsystem then had an optional backend that you can set to cmsis which just remapped these to the cmsis functions. Applications are then free to define their own backends. Would you like me to bring that back?

 https://github.com/zephyrproject-rtos/zephyr/pull/51059#discussion_r990924573

stephanosio commented:
Just to make it 100% clear:

For `arm_add_q31` declared as follows by CMSIS-DSP:

```
  /**
   * @brief Q31 vector addition.
   * @param[in]  pSrcA      points to the first input vector
   * @param[in]  pSrcB      points to the second input vector
   * @param[out] pDst       points to the output vector
   * @param[in]  blockSize  number of samples in each vector
   */
  void arm_add_q31(
  const q31_t * pSrcA,
  const q31_t * pSrcB,
        q31_t * pDst,
        uint32_t blockSize);
```

We want something like:

```
  /**
   * @brief Q31 vector addition.
   * @param[in]  src_a      points to the first input vector
   * @param[in]  src_b      points to the second input vector
   * @param[out] dst       points to the output vector
   * @param[in]  block_size  number of samples in each vector
   */
DSP_FUNC_SCOPE void zdsp_add_q31(
  const q31_t * src_a,
  const q31_t * src_b,
        q31_t * dst,
        uint32_t block_size);
```

i.e. same type system and names, "identical" argument names adjusted for different naming convention

Not:

```
/**
 * @brief q1_31 vector addition.
 * @param[in]  lhs      points to the first input vector
 * @param[in]  rhs      points to the second input vector
 * @param[out] dst      points to the output vector
 * @param[in]  count    number of samples in each vector
 */
DSP_FUNC_SCOPE void zdsp_add_q1_31(const q1_31_t *lhs, const q1_31_t *rhs, q1_31_t *dst,
				   uint32_t count);
```

i.e. no types like `q1_31_t` and `q16_16_t`, no different argument names (changes due to different naming convention is fine, but it should stay more or less the same -- e.g. no change from `block_size` to `count`),

yperess commented:
> Just to make it 100% clear:
> 
> For `arm_add_q31` declared as follows by CMSIS-DSP:
> 
> ```
>   /**
>    * @brief Q31 vector addition.
>    * @param[in]  pSrcA      points to the first input vector
>    * @param[in]  pSrcB      points to the second input vector
>    * @param[out] pDst       points to the output vector
>    * @param[in]  blockSize  number of samples in each vector
>    */
>   void arm_add_q31(
>   const q31_t * pSrcA,
>   const q31_t * pSrcB,
>         q31_t * pDst,
>         uint32_t blockSize);
> ```
> 
> We want something like:
> 
> ```
>   /**
>    * @brief Q31 vector addition.
>    * @param[in]  src_a      points to the first input vector
>    * @param[in]  src_b      points to the second input vector
>    * @param[out] dst       points to the output vector
>    * @param[in]  block_size  number of samples in each vector
>    */
> DSP_FUNC_SCOPE void zdsp_add_q31(
>   const q31_t *src_a,
>   const q31_t *src_b,
>         q31_t *dst,
>         uint32_t block_size);
> ```
> 
> i.e. same type system and names, "identical" argument names adjusted for different naming convention
> 
> Not:
> 
> ```
> /**
>  * @brief q1_31 vector addition.
>  * @param[in]  lhs      points to the first input vector
>  * @param[in]  rhs      points to the second input vector
>  * @param[out] dst      points to the output vector
>  * @param[in]  count    number of samples in each vector
>  */
> DSP_FUNC_SCOPE void zdsp_add_q1_31(const q1_31_t *lhs, const q1_31_t *rhs, q1_31_t *dst,
> 				   uint32_t count);
> ```
> 
> i.e. no types like `q1_31_t` and `q16_16_t`, no different argument names (changes due to different naming convention is fine, but it should stay more or less the same -- e.g. no change from `block_size` to `count`),

Updated this PR. I'll be adding the documentation tomorrow.

stephanosio commented on tests/lib/cmsis_dsp/basicmath/src/f32.c:
```diff
@@ -5,18 +5,18 @@
 #define REL_ERROR_THRESH	(5.0e-5)
 
-static void test_arm_add_f32(
+static void test_zdsp_add_f32(
```
This is very confusing -- a CMSIS-DSP test testing ZDSP?

With the introduction of the "Zephyr Math subsystem" (aka. ZDSP), there should be a dedicated testsuite for it (e.g. `tests/subsys/{math,dsp}/basicmath`).

I acknowledge that having both `tests/subsys/{math,dsp}/basicmath` and `tests/lib/cmsis_dsp/basicmath` is far from ideal since they are basically testing the same thing.

My suggestion is to move the `basicmath` (and other "migrated" API in the future) test to `tests/subsys/{math,dsp}/basicmath`.

p.s. should settle on either `math` or `dsp`.

stephanosio reviewed (changes requested):
Please update the commit message, and also split the changes into more manageable commits.

stephanosio commented on modules/Kconfig.cmsis_dsp:
```diff
@@ -29,6 +29,7 @@ config CMSIS_DSP_BASICMATH
 config CMSIS_DSP_COMPLEXMATH
 	bool "Complex Math Functions"
 	imply CMSIS_DSP_FASTMATH
+	depends on CPU_CORTEX
```
```suggestion
	depends on CPU_CORTEX || ARCH_POSIX
```

ruuddw commented:
This approach looks like something that could work for ARC as well. Maybe split it in two parts, certainly on commit level, but maybe on PR level as well:
- introduce the new dsp (I like that better than math) subsystem
- update the CMSIS-DSP to make it a backend

Let me check if the type definitions need some overrides for ARC (compiler may define them differently) and if the 'inline' mechanism works for an ARC backend with extra XY memory and AGU attributes.

yperess commented:
@stephanosio I'm seeing some errors (for example on `mm_feather`) that I'm not understanding how to debug. Any help would be appreciated:

```
[97/205] Building C object modules/cmsis/CMSIS/DSP/Source/BasicMathFunctions/CMakeFiles/..__modules__hal__cmsis__CMSIS__DSP__Source__BasicMathFunctions.dir/arm_and_u32.c.obj
FAILED: modules/cmsis/CMSIS/DSP/Source/BasicMathFunctions/CMakeFiles/..__modules__hal__cmsis__CMSIS__DSP__Source__BasicMathFunctions.dir/arm_and_u32.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DARM_MATH_NEON -DBOARD_FLASH_SIZE="CONFIG_FLASH_SIZE*1024" -DCPU_MIMXRT1062DVL6A -DKERNEL -DTC_RUNID=ae0feeea8321916043995ba0c848f9a4 -DXIP_BOOT_HEADER_DCD_ENABLE=1 -DXIP_BOOT_HEADER_ENABLE=1 -DXIP_EXTERNAL_FLASH -D_FORTIFY_SOURCE=2 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/workdir/zephyr/include -Izephyr/include/generated -I/workdir/zephyr/soc/arm/nxp_imx/rt -I/workdir/zephyr/lib/libc/newlib/include -I/workdir/zephyr/subsys/testsuite/include -I/workdir/zephyr/subsys/testsuite/ztest/include -I/workdir/modules/hal/cmsis/CMSIS/Core/Include -I/workdir/modules/hal/cmsis/CMSIS/DSP/Include -I/workdir/modules/hal/nxp/mcux/mcux-sdk/devices/MIMXRT1062 -I/workdir/modules/hal/nxp/mcux/mcux-sdk/devices/MIMXRT1062/drivers -I/workdir/modules/hal/nxp/mcux/mcux-sdk/devices/MIMXRT1062/xip -I/workdir/modules/hal/nxp/mcux/mcux-sdk/drivers/common -I/workdir/modules/hal/nxp/mcux/mcux-sdk/drivers/lpuart -I/workdir/modules/hal/nxp/mcux/mcux-sdk/drivers/igpio -I/workdir/modules/hal/nxp/mcux/mcux-sdk/drivers/cache/armv7-m7 -I/workdir/zephyr/modules/hal_nxp/. -I/workdir/zephyr/subsys/dsp/cmsis_backend/public_overrides -fno-strict-aliasing -Os -imacros /workdir/zephyr/twister-out/mm_feather/tests/subsys/dsp/basicmath/libraries.cmsis_dsp.basicmath.fpu/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m7 -mthumb -mabi=aapcs -mfpu=fpv5-d16 -mfloat-abi=hard -mfp16-format=ieee --sysroot=/opt/toolchains/zephyr-sdk-0.15.1/arm-zephyr-eabi/arm-zephyr-eabi -imacros /workdir/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -Werror -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/workdir/zephyr/tests/subsys/dsp/basicmath=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -specs=nano.specs -std=c99 -MD -MT modules/cmsis/CMSIS/DSP/Source/BasicMathFunctions/CMakeFiles/..__modules__hal__cmsis__CMSIS__DSP__Source__BasicMathFunctions.dir/arm_and_u32.c.obj -MF modules/cmsis/CMSIS/DSP/Source/BasicMathFunctions/CMakeFiles/..__modules__hal__cmsis__CMSIS__DSP__Source__BasicMathFunctions.dir/arm_and_u32.c.obj.d -o modules/cmsis/CMSIS/DSP/Source/BasicMathFunctions/CMakeFiles/..__modules__hal__cmsis__CMSIS__DSP__Source__BasicMathFunctions.dir/arm_and_u32.c.obj -c /workdir/modules/hal/cmsis/CMSIS/DSP/Source/BasicMathFunctions/arm_and_u32.c
In file included from /workdir/modules/hal/cmsis/CMSIS/DSP/Include/arm_math_types.h:91,
                 from /workdir/modules/hal/cmsis/CMSIS/DSP/Include/dsp/basic_math_functions.h:30,
                 from /workdir/modules/hal/cmsis/CMSIS/DSP/Source/BasicMathFunctions/arm_and_u32.c:29:
/opt/toolchains/zephyr-sdk-0.15.1/arm-zephyr-eabi/lib/gcc/arm-zephyr-eabi/12.1.0/include/arm_neon.h: In function 'arm_and_u32':
/opt/toolchains/zephyr-sdk-0.15.1/arm-zephyr-eabi/lib/gcc/arm-zephyr-eabi/12.1.0/include/arm_neon.h:11045:1: error: inlining failed in call to 'always_inline' 'vst1q_u32': target specific option mismatch
11045 | vst1q_u32 (uint32_t * __a, uint32x4_t __b)
      | ^~~~~~~~~
/workdir/modules/hal/cmsis/CMSIS/DSP/Source/BasicMathFunctions/arm_and_u32.c:100:9: note: called from here
  100 |         vst1q_u32(pDst, vandq_u32(vecA, vecB) );
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Commit:
dsp: move cmsis_dsp basicmath tests

Move the tests to tests/subsys/dsp/ as they'll be a part of the Zephyr
dsp subsystem.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
dsp: Update tests for basicmath to use dsp subsystem

Update the relevant tests in basicmath to use the dsp subsystem. Note
that f16 is not updated since it's much more architecture specific and
did not become a part of zdsp.

Signed-off-by: Yuval Peress <peress@google.com>

stephanosio commented on MAINTAINERS.yml:
```diff
@@ -327,8 +327,10 @@ CMSIS-DSP integration:
+    - yperess
   files:
     - modules/Kconfig.cmsis_dsp
+    - subsys/dsp/
```
There should be a separate area for "DSP subsystem" since, although the API is identical to that of the CMSIS-DSP, it is supposed a standalone math/DSP subsystem for Zephyr.

yperess replied:
Done

stephanosio commented on include/zephyr/math/dsp.h:
```diff
@@ -0,0 +1,74 @@
+ */
+
+/**
+ * @file zephyr/math/dsp.h
```
As others and I pointed out earlier, can we settle on either `math` OR `dsp`, instead of using these two terms interchangeably on an arbitrary basis?

I agree with @ruuddw and, IMHO, we should settle for `dsp` since the CMSIS-DSP API, after which this subsystem is modelled, is a "DSP" library (and not just generic "math" library).

such that there are `include/zephyr/dsp/dsp.h`, `include/zephyr/dsp/basicmath.h`, ...

yperess replied:
Done

stephanosio commented on subsys/dsp/Kconfig:
```diff
@@ -0,0 +1,14 @@
+	select DSP_BACKEND_HAS_STATIC
+	help
+	  Implement the various zephyr DSP functions using the CMSIS-DSP library. This feature
+	  requires the CMSIS module to be selected.
```
Please fix the "no new line at the end of the file."

yperess replied:
Done

stephanosio commented on subsys/dsp/cmsis_backend/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+
+target_include_directories(zdsp PUBLIC public_overrides)
+
+zephyr_compile_definitions_ifndef(CONFIG_ARM __GNUC_PYTHON__)
```
This really needs to be in the CMSIS-DSP integration code itself, not here in "DSP subsystem."

stephanosio commented on subsys/dsp/cmsis_backend/public_overrides/zdsp_backend/dsp.h:
```diff
@@ -0,0 +1,158 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef SUBSYS_MATH_CMSIS_BACKEND_PUBLIC_ZDSP_BACKEND_DSP_H_
+#define SUBSYS_MATH_CMSIS_BACKEND_PUBLIC_ZDSP_BACKEND_DSP_H_
```
Do we really need `subsys/dsp/cmsis_backend/public_overrides/zdsp_backend/dsp.h`?

Can't it be something simpler like: `subsys/dsp/cmsis/dsp_backend.h`?

stephanosio commented:
Renamed the PR title more appropriately.

carlescufi commented:
API meeting:

- @galak asks why this should be in `subsys/` instead of `lib/`
- The question of "what is a subsystem" is very blurry today
- @evgeniy-paltsev notes that ARC uses the CMSIS-DSP API, providing an ARC-optimized implementation
-

yperess commented:
> * @teburd asks whether the APIs are going to be guaranteed to behave in the same way for multiple architectures. @stephanosio mentions that the API should be generic enough to cover identical behavior for multiple architectures. @galak mentions rounding/precision errors in floating point

I added the documentation from CMSIS-DSP that includes the rounding information and what's expected from the implementation.

yperess commented:
> - @MaureenHelm mentions that enabling hardware accel on NXP platforms made the tests fail

Could you please provide details on how to replicate this? I'm not able to find the right configs.

MaureenHelm commented:
> > * @MaureenHelm mentions that enabling hardware accel on NXP platforms made the tests fail
> 
> Could you please provide details on how to replicate this? I'm not able to find the right configs.

#24655

teburd commented:
Some further questions I've had since the initial discussion..

Would types outside of those defined by cmsis-dsp be allowed? E.g. q1_23_t for audio dsps that have specialized 24bit MACs perhaps.

If other types are allowed, what is the expectation of this API, that a C fallback must always exist and behave the same way?

Can the dsp typedefs be provided separate from the functionality? E.g. I have my own DSP library I want to use with an API in Zephyr that uses these type aliases. So perhaps splitting out dsp_types.h from dsp_math.h would be useful. Currently they are pulled together in dsp.h

yperess commented:
> Some further questions I've had since the initial discussion..
> 
> Would types outside of those defined by cmsis-dsp be allowed? E.g. q1_23_t for audio dsps that have specialized 24bit MACs perhaps.
> 
> If other types are allowed, what is the expectation of this API, that a C fallback must always exist and behave the same way?
> 
> Can the dsp typedefs be provided separate from the functionality? E.g. I have my own DSP library I want to use with an API in Zephyr that uses these type aliases. So perhaps splitting out dsp_types.h from dsp_math.h would be useful. Currently they are pulled together in dsp.h
> 

Good point, I'll update

yperess commented:
> > > * @MaureenHelm mentions that enabling hardware accel on NXP platforms made the tests fail
> > 
> > 
> > Could you please provide details on how to replicate this? I'm not able to find the right configs.
> 
> #24655

I spent a few hours on this and I can't get PowerQuad working irrespective of this abstraction. I think this can be something we deal with when/if someone has the need to enable PowerQuad and is out of scope for this work.

yperess commented:
> Can the dsp typedefs be provided separate from the functionality? E.g. I have my own DSP library I want to use with an API in Zephyr that uses these type aliases. So perhaps splitting out dsp_types.h from dsp_math.h would be useful. Currently they are pulled together in dsp.h

Hey @teburd , I tried this and I'm not 100% sure I like it as much now. If I understand correctly, you want to define `q7_t` as something other than `int8_t` in your own library is that right? I'm not sure I see the benefit.

What I did was create a `zdsp_backend_types.h` which is included from `dsp.h` prior to any of the other headers. This allowed the backend to define `q7_t`, `q15_t`, `q31_t`, `float32_t` and `float64_t` on their own. I think if we do that, I would like to convert these from `typedef` to `#define` this way I can assert that they were all defined and give a more meaningful error in `dsp.h`

If I'm off, please correct me.

teburd commented:
I don’t want to redefine the types, I want the types to be available separate from the from the functions. So dsp_types.h gets included in dsp.h rather than dsp.h defining the types is all

yperess commented:
> I don’t want to redefine the types, I want the types to be available separate from the from the functions. So dsp_types.h gets included in dsp.h rather than dsp.h defining the types is all

Done, PTAL and let me know if this is what you wanted.

teburd commented:
> > I don’t want to redefine the types, I want the types to be available separate from the from the functions. So dsp_types.h gets included in dsp.h rather than dsp.h defining the types is all
> 
> Done, PTAL and let me know if this is what you wanted.

Yes, types.h makes sense. Thanks

yperess commented:
Ok, ready for review. All failures seem like false positives to me.

yperess commented:
Ping, any feedback on this? It's blocking a lot of other work.

andyross reviewed (dismissed):
I haven't been following the sensor evolution, but this seems mostly harmless to me.  As I'm reading it, it's a straightforward recapitulation of (a subset of) an existing industry API which can be fairly easily implemented anywhere (in fact as I read arm_math.h, it looks like there are generic C implementations already present?).

This seems probably worthwhile.  A fully generic "DSP math" library isn't going to be much use to anyone anyway.  Hardware varies widely in capabilities, these formats tend to overlap with hardware only sometimes, and they also tend to collide with compiler intrinsics for SIMD devices (I'm literally right now dealing with an emerging llvm port to Xtensa HiFi3/4/5 ISAs, for example).

Thought that said: it looks to me like ARM released this already in a portable form and a Zephyr-compatible license.  Would it not be possible to just rework CMSIS DSP into a Zephyr module for everyone?  Maybe not, but if so maybe that would be better still?

stephanosio commented:
> Would it not be possible to just rework CMSIS DSP into a Zephyr module for everyone?

FYI, It already is [integrated as a module](https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/Kconfig.cmsis_dsp). The purpose of this PR is to define Zephyr's own DSP API strictly based on the "CMSIS-DSP API," such that Zephyr applications do not directly depend on the CMSIS-DSP and other vendors can implement their own "backends" for it other than the CMSIS-DSP.

stephanosio commented on subsys/dsp/Kconfig:
```diff
@@ -0,0 +1,26 @@
+	  Implement the various zephyr DSP functions using the CMSIS-DSP library. This feature
+	  requires the CMSIS module to be selected.
+
+config DSP_BACKEND_THIRD_PARTY
+	bool "Do not use any Zephyr backends for DSP"
+	help
+	  Rely on the application to provide a DSP backend. The implementation should be added to
+	  the 'zdsp' build target by the application or one of its modules.
```
> The implementation should be added to the 'zdsp' build target by the application

This sounds more like a "first party" than a "third party."

I wonder if it would be better to name this `DSP_BACKEND_NONE`?

yperess replied:
I see what you mean, I'm OK with either name. I would think that `NONE` would mean it's also not needed, but in this case you'd get an error. So `THIRD_PARTY` was meant to say I'm choosing a non-Zephyr backend (I'm bringing in my own).

yperess replied:
@stephanosio I left this for now to give you a change to read my reasoning for the term `THIRD_PARTY`, let me know which way you want to go with it. I feel like `NONE` is misleading because the application would still need to provide one so maybe we can come up with a better term?

stephanosio replied:
On second thought, I agree `NONE` can be also quite misleading.

What do you think about `CUSTOM` then? It is a "custom backend" in the sense that it is not one of the backends provided by the DSP subsystem and user has to provide their own **custom** backend.

yperess replied:
Done

stephanosio commented on tests/subsys/dsp/basicmath/src/q31.c:
```diff
@@ -302,9 +302,9 @@ static void test_arm_shift_q31(
+DEFINE_TEST_VARIANT3(basic_math_q31, zdsp_shift_q31, possat, in_maxpos, ref_shift_possat, 9);
+DEFINE_TEST_VARIANT3(basic_math_q31, zdsp_shift_q31, negsat, in_maxneg, ref_shift_negsat, 9);
 
 static void test_arm_and_u32(
 	const uint32_t *input1, const uint32_t *input2, const uint32_t *ref, size_t length)
```
I see that this file still contains some `arm_` prefixed function (i.e. CMSIS-DSP function) tests; e.g. `test_arm_and_u32`.

This is very confusing and does not seem very logical -- since this is a **ZDSP** test, it should not contain any CMSIS-DSP tests.

Since splitting this test into CMSIS-DSP and ZDSP tests would be a lot of unnecessary work with little gain, why not just add `zdsp_and_u32` and its friends to the ZDSP API? After all, these integer DSP functions can be valuable since many SIMD instructions implement integer operations.

yperess replied:
Thanks, I missed those in my regex for swapping the tests over.

stephanosio commented:
@yperess https://github.com/zephyrproject-rtos/cmsis/pull/20 has been merged.

stephanosio commented on include/zephyr/dsp/basicmath.h:
```diff
@@ -0,0 +1,915 @@
+/* Copyright (c) 2022 Google LLC
```
should probably add ARM copyright here since this code is more or less directly based on their CMSIS-DSP headers.

yperess replied:
Done

fabiobaltieri commented on subsys/dsp/cmsis/public/zdsp_backend.h:
```diff
@@ -0,0 +1,162 @@
+DEF_ZDSP_MAP(sub_q31, q31_t)
+DEF_ZDSP_MAP(sub_f32, float32_t)
+
+#undef DEF_ZDSP_MAP
```
How about using more descriptive macro name rather than redefining the same one? I always found these patterns a bit harsh to read, could do `DEF_ZDSP_MAP_BASIC`, `DEF_ZDSP_MAP_SCALE`, etc...

yperess replied:
I went ahead and just expanded them, the macros were really useful to verify that this work, but I think it's more readable without them.

fabiobaltieri commented on tests/subsys/dsp/common/math_helper.h:
```diff
@@ -0,0 +1,276 @@
+/* ----------------------------------------------------------------------
```
Would it make sense to only keep a copy of these and `test_common.h`?

yperess replied:
Done, I kept just one copy and used target_include_directories in the new test location.

SiyuanCheng-CN commented:
> This approach looks like something that could work for ARC as well. Maybe split it in two parts, certainly on commit level, but maybe on PR level as well:
> 
> * introduce the new dsp (I like that better than math) subsystem
> * update the CMSIS-DSP to make it a backend
> 
> Let me check if the type definitions need some overrides for ARC (compiler may define them differently) and if the 'inline' mechanism works for an ARC backend with extra XY memory and AGU attributes.

@yperess It is verified that the proposed API works well for ARC and can be mapped to ARC-backend dsp library. To map with ARC dsp library, `__agu` and` __restrict` are needed to be added for every src and dst input variables of API in `basicmath.h` and` zdsp_backend.h`. You can define two macros like `#define ARC_AGU __agu`, `#define ARC_RESTRICT __restrict` when using ARC-backend dsp library. API in `basicmath.h` should look like 
`DSP_FUNC_SCOPE void zdsp_mult_q7(const ARC_AGU q7_t * ARC_RESTRICT src_a, const ARC_AGU q7_t * ARC_RESTRICT src_b, ARC_AGU q7_t * ARC_RESTRICT dst, uint32_t block_size);`

teburd commented:
The macro adds could be done in a subsequent PR and should be named genetically.

yperess commented:
@teburd @fabiobaltieri @stephanosio force push changes:
- I got rid of the macros in zdsp_backend.h that generated the static inline functions, should look cleaner now
- I updated the Kconfig `DSP_BACKEND_THIRD_PARTY` to be `DSP_BACKEND_CUSTOM`
- I removed the duplicate test `common` directory

ruuddw commented:
> The macro adds could be done in a subsequent PR and should be named generically.

Yes, in the generic interface there should not be _ARC_. The [restricted pointers](https://gcc.gnu.org/onlinedocs/gcc/Restricted-Pointers.html), however, are a generic feature also supported by GCC and other compilers.
Since we're defining a new public interface, I'd rather see it added including the attribute placeholders from the beginning, instead of adding them later. How about adding something like "DSP_ARGP" for the pointer arguments in the basicmath.h functions? By default it could be mapped to __restricted__ or left empty.

yperess commented:
> > This approach looks like something that could work for ARC as well. Maybe split it in two parts, certainly on commit level, but maybe on PR level as well:
> > 
> > * introduce the new dsp (I like that better than math) subsystem
> > * update the CMSIS-DSP to make it a backend
> > 
> > Let me check if the type definitions need some overrides for ARC (compiler may define them differently) and if the 'inline' mechanism works for an ARC backend with extra XY memory and AGU attributes.
> 
> @yperess It is verified that the proposed API works well for ARC and can be mapped to ARC-backend dsp library. To map with ARC dsp library, `__agu` and` __restrict` are needed to be added for every src and dst input variables of API in `basicmath.h` and` zdsp_backend.h`. You can define two macros like `#define ARC_AGU __agu`, `#define ARC_RESTRICT __restrict` when using ARC-backend dsp library. API in `basicmath.h` should look like `DSP_FUNC_SCOPE void zdsp_mult_q7(const ARC_AGU q7_t * ARC_RESTRICT src_a, const ARC_AGU q7_t * ARC_RESTRICT src_b, ARC_AGU q7_t * ARC_RESTRICT dst, uint32_t block_size);`

Do the positions of these arguments matter? Does `__agu` always need to come before the type and `__restrict` before the argument?

stephanosio commented:
@ruuddw Do the attributes like `__agu` need to be specified for the function declarations? If not, it would be preferable to leave the declarations alone and only specify them in the function definitions (i.e. in the body).

ruuddw commented:
> My understanding is that `__restrict` will break this behavior.

Correct, when using __restrict the assumption/precondition is that there is no overlap between the inputs and outputs. As the compiler may do SW pipelining and reorder/schedule load and store instructions in an optimal way, as in this example:

```
for(i = 0; i < 20; i++)
   *a++ = *b++;
}
```
Without restricted pointers:
```
ld.ab %r2,[%r1,8]
st.ab %r2,[%r0,8] ; Stall for LD to complete
ld %r2,[%r1,-4]
st %r2,[%r0,-4] ; Stall for LD to complete
```
With restricted pointers
```
ld.ab %r3,[%r1,8]
ld %r2,[%r1,-4]
st.ab %r3,[%r0,8] ; Notice compiler uses R2 and R3; interleaving accesses
st %r2,[%r0,-4]
```

Looking at the definitions of the basicmath functions, they seem to imply that inputs cannot be modified/overwritten as they have the 'const' qualifier (though that's not the C semantics as aliased pointers are not forbidden or prevented by const.).

For a generic Zephyr API, I'd be ok to take the strict definition: when output is overlapping input buffers, the result is not guaranteed. That leaves freedom to implementations, and works both with backends that allow overwriting inputs and that don't.

SiyuanCheng-CN commented:
@yperess 

> Does __agu always need to come before the type and __restrict before the argument?
Yes, and you can take #52478 as an example, which is a draft PR for ARC dsplib backend test.

> I couldn't find any information about __agu
This is a specific attribute in MetaWare compiler, which is used to denote pointers that need to use AGU feature.

yperess commented:
> @yperess 
> 
> > Does __agu always need to come before the type and __restrict before the argument?
> 
> Yes, and you can take #52478 as an example, which is a draft PR for ARC dsplib backend test.
> 
> > I couldn't find any information about __agu
> 
> This is a specific attribute in MetaWare compiler, which is used to denote pointers that need to use AGU feature.

Is there an existing Kconfig for the AGU (I couldn't find one)? Something like `config ARCH_ARC_AGU` that we can leverage? So we know what other architectures support an AGU?

yperess commented:
@teburd @stephanosio @SiyuanCheng-CN @ruuddw it seems like the discussion about flags is going to take a while to figure out. Thoughts about merging this as is so it can unblock my sensor work while we discuss?

stephanosio commented:
> @teburd @stephanosio @SiyuanCheng-CN @ruuddw it seems like the discussion about flags is going to take a while to figure out. Thoughts about merging this as is so it can unblock my sensor work while we discuss?

I agree. The ARC-specific function attribute handling details can be finalised later and implemented as an enhancement to the DSP subsystem.

SiyuanCheng-CN commented:
@yperess 

> Is there an existing Kconfig for the AGU

Here is the [kconfig link](https://github.com/zephyrproject-rtos/zephyr/blob/672c67f1c6841e87ab5f9b08a378d9dc06385e2e/arch/arc/core/dsp/Kconfig) related to AGU. It's an ARC-specific feature, which eliminates the need for separate load/store instructions during DSP operation.
I agree with adding it in the subsequent PR, and it is already in a draft.

stephanosio commented on tests/subsys/dsp/basicmath/src/f16.c:
```diff
@@ -9,7 +9,7 @@
-#include "../../common/test_common.h"
+#include "common/test_common.h"
 
 #include "f16.pat"
```
This file still contains `arm_` prefixed function (i.e. CMSIS-DSP) tests in spite of being part of a **ZDSP** test.

There needs to exist logical separation between CMSIS-DSP and ZDSP and it does not make sense for a ZDSP test to directly test CMSIS-DSP functions.

stephanosio replied:
> Note that f16 is not updated since it's much more architecture specific and did not become a part of zdsp.

`_Float16` is available as part of ISO/IEC TS 18661-3, which is an extension to the ISO/IEC 9899 (i.e. ISO C standard). It would make sense for the ZDSP subsystem to support 16-bit float type.

yperess replied:
Done

ruuddw commented:
I'm ok with handling the extensions for ARC separately (see #52478), but we'll have to resolve the discussions within this release to avoid introducing an API that gets modified in a next release again.

I asked about the __restrict and input buffer re-use with our DSP experts. Despite the semantics of __restrict, it doesn't mean that things are guaranteed to be broken when input buffers are reused for outputs. For simple 'a = a + b' cases there should not be a problem even with __restrict. Typically whether this is allowed or not is determined more by the algorithm nature and C code then by the compiler handling __restrict optimizations.

yperess commented:
> I'm ok with handling the extensions for ARC separately (see #52478), but we'll have to resolve the discussions within this release to avoid introducing an API that gets modified in a next release again.
> 
> I asked about the __restrict and input buffer re-use with our DSP experts. Despite the semantics of __restrict, it doesn't mean that things are guaranteed to be broken when input buffers are reused for outputs. For simple 'a = a + b' cases there should not be a problem even with __restrict. Typically whether this is allowed or not is determined more by the algorithm nature and C code then by the compiler handling __restrict optimizations.

I don't know, I'll need to test a few things. Here's a great example of the optimization that can be done with `restrict` https://en.m.wikipedia.org/wiki/Restrict#Optimization. To me it seems like adding that attribute is risky, but I'll let others weigh in.

yperess commented:
@ruuddw @SiyuanCheng-CN @teburd @stephanosio 

Upon further review and consulting with some of our compiler guys I've reached the following conclusions:

### RE: `__restrict`

We should not allow `__restrict` as a pointer attribute in this API. By definition it means that the behavior of `a += b` is undefined and can break with different compilers or even versions of the same compiler.

### RE: `__agu`

This seems perfectly reasonable, though I would like it to go in a separate CL so we can have a focused discussion on how are we going to add architecture specific attributes to this API (which I assume we'll be running into quite a lot and this will be a long discussion that I don't want my other work to be blocked on).

ruuddw commented:
I agree with the __restrict 'risk' and the AGU is obviously not relevant for all architectures/implementations. Therefore the proposal is to add these via macro's that user can configure/define: if not applicable for a certain implementation backend, the preprocessor removes them.
The main discussion I think that needs to be resolved (independent of interface implementations) is the semantics that we want. Should a = math_op(a,b) and b = math_op(a,b) both be supported by any implementation backend? Or only reuse of the first (or second)? Should it be supported for any math_op, or only for simpler ones?
My preference would be to have the interface semantics defined more strict (i.e. adhering to our optimized __restrict), but if the majority prefers the semantics to be defined including in-place/argument aliasing support we will find other ways to have optimal DSP performance and compatibility with the ARC dsp library for our users.

yperess commented:
> I agree with the __restrict 'risk' and the AGU is obviously not relevant for all architectures/implementations. Therefore the proposal is to add these via macro's that user can configure/define: if not applicable for a certain implementation backend, the preprocessor removes them. The main discussion I think that needs to be resolved (independent of interface implementations) is the semantics that we want. Should a = math_op(a,b) and b = math_op(a,b) both be supported by any implementation backend? Or only reuse of the first (or second)? Should it be supported for any math_op, or only for simpler ones? My preference would be to have the interface semantics defined more strict (i.e. adhering to our optimized __restrict), but if the majority prefers the semantics to be defined including in-place/argument aliasing support we will find other ways to have optimal DSP performance and compatibility with the ARC dsp library for our users.

That seems fair, I guess the tradeoff we're looking at is:

### Option 1

Disallow in-place DSP operations. Which will allow for `__restrict` and generally require more memory in places where in-place arithmetic would have been used.

### Option 2

Disallow `__restrict` optimization thus making in-place operations possible (potentially saving some memory) at the cost of some performance.

ruuddw commented:
Nice summary of the options, that's indeed what I think needs to be decided. For a generic API, my vote is to have the stricter semantics. But we probably should get some more inputs on the user side: is there a lot of application code that assumes in-place processing is supported? I'm also not sure if CMSIS-DSP defines this explicitly.

yperess commented:
> Nice summary of the options, that's indeed what I think needs to be decided. For a generic API, my vote is to have the stricter semantics. But we probably should get some more inputs on the user side: is there a lot of application code that assumes in-place processing is supported? I'm also not sure if CMSIS-DSP defines this explicitly.

I posted this last night, I think there's a third option for a Kconfig, but it makes things a bit more complex.
https://github.com/zephyrproject-rtos/zephyr/discussions/52617

teburd reviewed (dismissed):
Looks good, should add the experimental status to the API overview.rst and a minimal subsys doc that gives a brief overview of the intent here.

https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/api/overview.rst

In my mind this is fine for experimental status, with the expectation that some things may change, potentially in breaking ways still.

stephanosio commented on tests/subsys/dsp/basicmath/testcase.yaml:
```diff
@@ -1,6 +1,6 @@
 tests:
   libraries.cmsis_dsp.basicmath:
```
These are no longer CMSIS-DSP tests, so the test names need to be updated as well:

```suggestion
  dsp.basicmath:
```

yperess replied:
Done

stephanosio commented on MAINTAINERS.yml:
```diff
@@ -321,6 +321,16 @@ CMSIS API layer:
 
+DSP subsystem:
+  status: maintained
+  maintainers:
+    - yperess
```
Please add me as a maintainer as well. I would be interested in any changes to the CMSIS-DSP backend and the ZDSP tests that have been relocated from the CMSIS-DSP tests.

yperess replied:
Done

stephanosio reviewed (changes requested):
Looks almost ready, just a minor change for the test names.

stephanosio commented on include/zephyr/dsp/types.h:
```diff
@@ -0,0 +1,71 @@
+ * @typedef float16_t
+ * @brief 16-bit floating point type definition.
+ */
+#if defined(CONFIG_FP16)
+typedef __fp16 float16_t;
```
Just a note to myself (or anyone who might be interested in maintaining this in the future) that `CONFIG_FP16` is currently an ARM-specific config and `__fp16` is an ARM-specific type.

The `CONFIG_FP16` Kconfig should eventually be moved to the same level as `CONFIG_FPU` and the architecture-agnostic ISO `_Float16` type support should be implemented.

ruuddw replied:
Without knowing the details myself: is there a reason for not using the ISO _Float16 already now then? Is it semantically different from __fp16?

stephanosio replied:
On ARM, the `__fp16` type can represent a half-precision floating point value in either IEEE 754-2008 format or the "ARM alternative format" depending on the specified compiler options -- we do not want to take away that capability for now.

stephanosio replied:
https://gcc.gnu.org/onlinedocs/gcc/Half-Precision.html

teburd commented:
A few compliance issues look legit

stephanosio commented on doc/services/dsp/index.rst:
```diff
@@ -0,0 +1,61 @@
+.. _dsp_api:
```
This needs to match the ref in `overview.rst`.

```suggestion
.. _zdsp_api:
```

yperess replied:
Odd that it passed the docs build for me locally. Thanks for catching this.

teburd commented on modules/Kconfig.cmsis:
```diff
@@ -22,7 +22,7 @@ endif
 menuconfig CMSIS_DSP
 	bool "CMSIS-DSP Library Support"
-	depends on (CPU_CORTEX && NEWLIB_LIBC) || ARCH_POSIX
+	depends on NEWLIB_LIBC || ARCH_POSIX
```
Not a blocker, but a note here is that depends on NEWLIBC should be revisited when PICOLIBC becomes the thing
